### DISCUSSION TITLE
RFC: Pattern Matching

### DISCUSSION THREAD

**Discussion by flavorjones** [sparklemotion/nokogiri#2360] [RFCs]
## Pattern Matching

**If Nokogiri implemented a pattern matching interface for `Nokogiri::XML::Node` and `Nokogiri::XML::NodeSet`, what should it look like?**

If you have some code you're using to deconstruct Nokogiri objects, please post a comment pointing us to what you've done, and explain a bit about your use case and why that approach is valuable, if you can.

If you have use cases for which XPath or CSS queries are awkward or slow, please post a comment explaining what you'd like to be able to do.

## Resources

@baweaver has been very thoughtful on the topic of general pattern matching support, he's written [Pattern Matching Interfaces in Ruby - Google Docs](https://docs.google.com/document/d/1spnuQTKy5i7Lx-sDCsORKN981Iam1IuNdOsYkhh9Yi0/edit#heading=h.vmv48c5zz9sr) to start a conversation around basic principles of what "good" looks like in a PM interface

He's also suggested something like this at https://twitter.com/keystonelemur/status/1357424192904454145:

```ruby
# Pattern to match a link: <a/>  
node in { name: 'a' }

# Pattern for a link with a class: <a class="foo"/>
node in { name: 'a', class: 'foo' }

# Pattern to select that link: <a class="foo">{.}</a>
node => { name: 'a', class: 'foo', children: }

# Pattern to select the target URL: <a class="foo" href="{.}"/>
node => { name: 'a', href: }

# Pattern for multiple links: <a class="foo" href="{.}"/>+
nodes => [*, { name: 'a', href: => link_one }, { name: 'a', href: => link_two } *]
```

**Comment by soulcutter** [comment]
I imagine this sort of interface being valuable for matching nodes while parsing streams - XPath is clunky for that sort of thing, and SAX is too low-level for most people to want to deal with it.

This might give me a reason to dust off my old sax wrapper https://github.com/soulcutter/saxerator

**Comment by flavorjones** [comment]
I've published a PR with an experimental API at https://github.com/sparklemotion/nokogiri/pull/2523, would love feedback on that pull request from anybody with an opinion.

**Comment by flavorjones** [comment]
Just a bump for anyone following. #2523 has received a bit more love and has some pattern matching use cases in the test suite. Would love to hear any feedback anybody has on this. Likely this will land as "experimental" in v1.14.0.

**Comment by flavorjones** [comment]
Looks like this was useful to someone! https://github.com/rails/rails/pull/49003

Will document pattern matching as no-longer-experimental in the next minor release.
